core/meta/src/TClass.cxx

```diff
@@ -3455,38 +3455,32 @@ const char *TClass::GetSharedLibs()
 
 TList *TClass::GetListOfBases()
 {
-   if (!fBase) {
-      if (fCanLoadClassInfo) {
-         if (fState == kHasTClassInit) {
+   if (fBase)
+      return fBase;
 
-            R__LOCKGUARD(gInterpreterMutex);
-            // NOTE: Add test to prevent redo if another thread has already done the work.
-            // if (!fHasRootPcmInfo) {
+   if (fState == kHasTClassInit) {
+      R__LOCKGUARD(gInterpreterMutex);
 
-            // The bases are in our ProtoClass; we don't need the class info.
-            TProtoClass *proto = TClassTable::GetProtoNorm(GetName());
-            if (proto && proto->FillTClass(this)) {
-               // Not sure this code is still needed
```

Comment:
This comment is being removed by the commit but it is not explain why?  i.e. Do we now understand why this code path is useful?

Reply:
I assumed the code comment was about `R__ASSERT(kFalse)`, not this code path in general. The truth, however, is that whoever added this comment should have "been sure" about if this code is needed or not and either left it there or removed it. This comment confuses more than it helps, that's why I removed it.

Reply:
> This comment confuses more than it helps, that's why I removed it.

On those kind of case, you still need to add a comment (at least in the commit) to explaning why the comment and its commented out assert are not needed.   The original commenting was obviously unsure whether this code path was or was not being used ...

Reply:
Another point of confusion about this code is that fHasRootPcmInfo is supposed to be true if there is a valid TProtoClass for that class ... so why does one need to set it to true here ... Answering this question is seemingly non trivial ....

Reply:
Well, I don't know. I only found out that the lock around the whole function is a performance issue that needs to be addressed. Since this code is tricky, I think that I will leave this one for you or Axel to figure out later.

Reply:
>  I only found out that the lock around the whole function is a performance issue that needs to be addressed.

I must be reading the (new) code incorrectly, I do not see (yet?) how the new code reduce the lock cost ... Can you clarify?

Reply:
And what I mean by a performance issue is that even when fBase has already been computed, threads keep taking the lock instead of just returning the already computed value. Since `GetListOfBases()` is called often, threads keep stepping on each other's toes all the time because of this. Please have a look at the screenshot attached in [ROOT-8876](https://sft.its.cern.ch/jira/browse/ROOT-8876) for more information.

Reply:
> when fBase has already been computed, threads keep taking the lock instead of just returning the already computed value.

I don't understand how.  Isn't the existing code
```
if ( fBase == nullptr) { .... }
return fBase;
```
?

Reply:
I don't know, maybe the branch predictor tries to run the body, and before `fCanLoadClassInfo` was frequently true, which meant taking a lock. Thinking about the code after the change, this problem may not be there anymore. In any case, the body of the if is large, so maybe the performance was due to the frequent long jump instead of taking locks. I did see improvement after my changes. I will rerun some of my tests to make sure this is not a problem in any case.

Reply:
However, I really think that adding whatever logic is needed to create the list of bases into `TCling::CreateListOfBaseClasses()` and making the code as I suggest below is probably not a bad idea. Currently the complexity is spread over several files, which makes it hard to reason about anything.

Reply:
Can you also try with lines 3467-3470 of your new code being
```
              fHasRootPcmInfo = kTRUE;
   
     if (fBase)
        return fBase;

    if (fCanLoadClassInfo)
```

Reply:
i.e. it is plausible that the "force to load the header when the list of base class is first requested" might have a positive effect on run-time on your case.  (As this seems to be the only functional (beside the longjump becoming a shortjump) change I can detect in reading the code.